### PR TITLE
Fix slice unpad

### PR DESCRIPTION
**Pull request by ASKabalan**
Fixed slice_unpad, now it supports complex numbers

### CODE REVIEW

**Comment by ASKabalan** [comment]
Ignore the formatting error.
It is fixed in another PR
Once you merge https://github.com/DifferentiableUniverseInitiative/jaxDecomp/pull/17 and https://github.com/DifferentiableUniverseInitiative/jaxDecomp/pull/12 and I wll rebase and this should be ok

File: tests/test_padding.py [hunk 1]
```diff
@@ -25,19 +26,35 @@
...
-      ],
-      key=jax.random.PRNGKey(rank))
+  if complex:
+
+    local_array = jax.random.normal(
+        shape=[
+            global_shape[0] // pdims[1], global_shape[1] // pdims[0],
+            global_shape[2]
+        ],
+        key=jax.random.PRNGKey(rank)) + 1j * jax.random.normal(
```
Comment (aboucaud) [inline tests/test_padding.py:43]: This line is an abomination :) 
First read I thought you were using complex keys `¯\_(ツ)_/¯` 

Instead I would leave the `local_array` from before untouched and just add a complex array to it in the `if complex:` statement

File: tests/test_padding.py [hunk 2]
```diff
@@ -25,19 +26,35 @@
...
+            shape=[
+                global_shape[0] // pdims[1], global_shape[1] // pdims[0],
+                global_shape[2]
+            ],
+            key=jax.random.PRNGKey(rank + 1))
```
Comment (aboucaud) [inline tests/test_padding.py:48]: I know this is just for tests, but I imagine that since the rank is the process index, this offset should be bigger than this, or you could just split the previous key.

**Comment by EiffL** [comment]
@ASKabalan can you try to rerun your precommit locally to squash the formatting test error?

**Comment by ASKabalan** [comment]
Yes all is green locally.
But at some point this formatting error correction slipped [here](https://github.com/DifferentiableUniverseInitiative/jaxDecomp/pull/12/commits/9d0c5398687d56e64c7d136a03fbb6bd665aa02a) and I think it fixes this issue.
Any ways this PR should come after the other two

**Comment by ASKabalan** [comment]
rebased and fixed commit history

File: jaxdecomp/_src/padding.py [hunk 3]
```diff
@@ -190,16 +189,11 @@ def per_shard_impl(arr: ArrayLike, padding_width: int | tuple[int]) -> Array:
...
-        arr,
-        padding_value=0.0,
-        padding_config=((-first_x, -last_x, interiour_padding),
-                        (-first_y, -last_y, interiour_padding),
-                        (-first_z, -last_z, interiour_padding)))
+    last_x = arr.shape[0] - last_x
+    last_y = arr.shape[1] - last_y
+    last_z = arr.shape[2] - last_z
+
+    return arr[first_x:last_x, first_y:last_y, first_z:last_z]
```
Comment (aboucaud) [inline jaxdecomp/_src/padding.py:196]: I guess this could be simplified to 
```suggestion
    return arr[first_x:-last_x, first_y:-last_y, first_z:-last_z]
```
but your suggestion is still clear

Comment (ASKabalan) [reply jaxdecomp/_src/padding.py:196]: Ok I remember why I did this
I had your version before and I had an issue when halo_size is 0 (no halo for a certain axis)
obviously I can have if branches to apply or not apply the unpadding, but I found that this way is clearer
do you agree?
